### PR TITLE
Food Morale recalculation, Taste Modifier Bionic fixes

### PR DESCRIPTION
**Pull request by KheirFerrum**
#### Summary

SUMMARY: Bugfixes "Redo morale calculation for food and fix Taste Modifier CBM power costs."

#### Purpose of change

Morale Calculation math for food is weird, and due to how Taste modifier CBM only calculates based on the food's base enjoyability this leads to wrong costs in certain situations.

#### Describe the solution

Forces all rotten food to use a minimum of -10 morale malus unless their original malus + 5 is higher (dog/cat food).  This prevents quirky situations like rotten Glazed Tenderloin still providing 4 morale on consumption. ~~Placing this on a separate if clause from Lupine/Feline mutant stuff means that Lupine/Feline mutants no longer gain _more_ morale from rotting pet food.~~ Lupine/Feline mutant stuff is placed earlier but still subject to the appropriate rotting modifier, so Lupine/Feline mutants will hate life less when they eat rotting pet food compared to a human who has to eat it.

Refactored calculations. Lupine and Feline mutants are no longer able to dodge monotony debuffs by eating dog/cat food. Gourmand now only affects eating unrotted food, while the Taste Blocker CBMs now are able to block _all_ negative morale modifiers, not just the stuff that starts negative.

Reworked power consumption of Taste Blocker CBMs, previously it only based it off the food's base enjoyability, but this can change if the food is rotted, or it is lupine/feline food and you are a lupine/feline mutant. Now power consumption is based off the morale debuff that it counters. ~To make this work smoothly a new argument was added to `Character::for_fun` called `bioblock`, by default it is true, when set to false it will ignore the effects of the Taste Modifier CBM.~ The morale changes were applied directly to `Character::modify_morale()`. To make the display congruent an additional change was made to `game_inventory.cpp` so that if the bionic is active and there is enough power negative morale will be displayed in brackets and coloured gray.

Edited the tests so that they will account for the new system. Added a new test for dog/cat food that will test if rotten dog/cat food is still unpalatable to Lupine/Feline mutants.

#### Describe alternatives you've considered

- Make Gourmand apply to rotten food.
I could, but felt that this was unnecessary, Gourmand already provides more than a few bonuses, and at the end of the day you're just a big eater, you're not going to suddenly enjoy rotten food. Furthermore, removing the morale penalty doesn't stop the fact that you will get food poisoning from eating rotting meals.

#### Testing

Spawned dog food, spawned glazed tenderloin. Checked that eating dog food with Taste Modifier on did not provide morale malus and used up energy appropriately. Checked that Lupine threshold mutation allowed eating of dog food and granted morale buffs, and no consumption of power from Taste Modifier CBM. Waited several days for dog food and glazed tenderloin to spoil, ate both and checked that it consumed power even with the Lupine threshold mutation.

#### Additional context

Monotony is currently a massive penalty that needs to be revised. In the future a food category system with subsets for snacks would be good. Otherwise a quick rebalance of the penalties themselves could act as a stopgap measure.
image

### CODE REVIEW

**Comment by Coolthulhu** [comment]
>```
> /home/runner/work/Cataclysm-BN/Cataclysm-BN/src/consumption.cpp:521:21: error: redundant boolean literal supplied to boolean operator [readability-simplify-boolean-expr,-warnings-as-errors]
>    if( bioblock == true && fun < 0 && has_active_bionic( bio_taste_blocker ) &&
>        ~~~~~~~~~~~~^~~~
>        bioblock
> ```

**Comment by Coolthulhu** [comment]
It's hard to read it like this. Shows why `bool` arguments are a bad practice.
Instead, it would be better to leave the negative morale being visible, just apply/prevent it in `modify_morale`.
Negative fun is good info even with taste blocker, since blocker's power drain depends on it.

**Comment by KheirFerrum** [comment]
> It's hard to read it like this. Shows why `bool` arguments are a bad practice. Instead, it would be better to leave the negative morale being visible, just apply/prevent it in `modify_morale`. Negative fun is good info even with taste blocker, since blocker's power drain depends on it.

I was mostly wondering where to inject the Taste Blocker and similar effects so that the cost math is congruent. I'll look at it later this week.

**Comment by KheirFerrum** [comment]
Test case had to be updated to check morale changes instead. May want to add a test case for rotten dog/cat food and lupine/feline mutation thresholds. Mainly to ensure that when the dog/cat food is rotten the mutants don't somehow gain morale due to weird code.

**Comment by Coolthulhu** [comment]
Now it looks pretty clean and nice.

File: tests/food_fun_for_test.cpp [hunk 1]
```diff
@@ -280,7 +302,7 @@ TEST_CASE( "fun for bionic bio taste blocker", "[fun_for][food][bionic]" )
 
                 THEN( "the bad taste is nullified" ) {
                     dummy.eat( garlic );
-                    CHECK( dummy.get_morale( MORALE_FOOD_BAD ) == 0 );
+                    CHECK( !dummy.get_morale( MORALE_FOOD_BAD ) );
```
Comment (Coolthulhu) [inline tests/food_fun_for_test.cpp:305]: I prefer explicit zero comparisons in cases like this.
It's common knowledge that `!0` is `true`, but zero comparison is more explicit.

**Pull request merged**